analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/SymbolicValue.cs

```diff
@@ -74,5 +72,62 @@ public sealed record SymbolicValue
             Constraints.Any()
                 ? Constraints.Values.Select(x => x.ToString()).OrderBy(x => x).JoinStr(", ")
                 : "No constraints";
+
+        private static SymbolicValue AddOrReplaceConstraint(SymbolicValue baseValue, SymbolicConstraint constraint)
+        {
+            if (baseValue.HasConstraint(constraint))
+            {
+                return baseValue;
+            }
+
+            var constraintCount = baseValue.Constraints.Count;
+            if (constraintCount == 0)
+            {
+                return GetOrAddSingleConstraint(baseValue, constraint);
+            }
+
+            if (constraintCount == 1 && baseValue.Constraints.ContainsKey(constraint.GetType()))
+            {
+                return AddOrReplaceConstraint(Constraintless, constraint);
+            }
+
+            return baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) };
+        }
+
+        private static SymbolicValue RemoveConstraint(SymbolicValue baseValue, SymbolicConstraint constraint) =>
+            baseValue.HasConstraint(constraint) ? RemoveConstraint(baseValue, constraint.GetType()) : baseValue;
+
+        private static SymbolicValue RemoveConstraint<T>(SymbolicValue baseValue) where T : SymbolicConstraint =>
+            baseValue.HasConstraint<T>() ? RemoveConstraint(baseValue, typeof(T)) : baseValue;
+
+        private static SymbolicValue RemoveConstraint(SymbolicValue baseValue, Type type)
+        {
+            var constraintCount = baseValue.Constraints.Count;
+            if (constraintCount == 1)
+            {
+                return Constraintless;
+            }
+
+            if (constraintCount == 2)
+            {
+                var otherConstraintType = baseValue.Constraints.Keys.FirstOrDefault(x => x != type);
+                return GetOrAddSingleConstraint(Constraintless, baseValue.Constraints[otherConstraintType]);
+            }
+
+            return baseValue with { Constraints = baseValue.Constraints.Remove(type) };
+        }
+
+        private static SymbolicValue GetOrAddSingleConstraint(SymbolicValue baseValue, SymbolicConstraint constraint)
+        {
+            if (singleConstraintCache.TryGetValue(constraint.Kind, out var result))
+            {
+                return result;
+            }
+            else
+            {
+                result = baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) };
+                return singleConstraintCache.GetOrAdd(constraint.Kind, result);
+            }
```

Comment:
We can not use `singleConstraintCache.GetOrAdd` directly without creating `baseValue with { Constraints = baseValue.Constraints.SetItem(constraint.GetType(), constraint) }` (which is totally against the idea of the cache) or passing a lambda factory. The lambda itself also allocates a delegate (and potentially a capture class) in the background so we shouldn't use one either. This implementation guarantees an allocation-free cache lookup.

Reply:
Now I've came back to this and realized I proposed the lamba... The capturing can be solved with static lambda, right? Othwerwise lambda itself is fine.

Reply:
> The capturing can be solved with static lambda, right?

We can not use a static lambda because we need to capture `baseValue`. I kept the implementation as is, because we want to be allocation free in cases of cache hits. Otherwise, we would ruin all the effort.